File: app/src/main/java/com/berkatfaatulohalawa1711010164/facevoting/fragment/akun_fragment.java
package com.berkatfaatulohalawa1711010164.facevoting.fragment;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.berkatfaatulohalawa1711010164.facevoting.API.APIService;
import com.berkatfaatulohalawa1711010164.facevoting.API.NoConnectivityException;
import com.berkatfaatulohalawa1711010164.facevoting.R;
import com.berkatfaatulohalawa1711010164.facevoting.config.Constants;
import com.berkatfaatulohalawa1711010164.facevoting.helper.ErrorHelper;
import com.berkatfaatulohalawa1711010164.facevoting.helper.SessionHelper;
import com.berkatfaatulohalawa1711010164.facevoting.model.AkunModel;
import com.berkatfaatulohalawa1711010164.facevoting.model.ErrorModel;
import com.berkatfaatulohalawa1711010164.facevoting.ui.Login;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.internal.EverythingIsNonNull;


public class akun_fragment extends Fragment {
    private ProgressDialog pDialog;
    private EditText mEmail,mNama,mNik,mPassword;
    private Button mEdit,mLogout;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if(getActivity() != null){
            getActivity().setTitle("Setting");
        }
        SharedPreferences sharedPreferences = getActivity().getSharedPreferences(
                Constants.USER_KEY, Context.MODE_PRIVATE);
        String idUser = sharedPreferences.getString("id_user", "");
        View myview = inflater.inflate(R.layout.fragment_akun, container, false);
        setData(myview);
        loadData(idUser);
        mEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateUser(idUser);
            }
        });
        mLogout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SessionHelper.logout(getContext());
                Intent intent = new Intent(getActivity(), Login.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                startActivity(intent);
                if(getActivity()!= null){
                    getActivity().finish();
                }
            }
        });
        return myview;
    }

    private void updateUser(String idUser){
        String nama = mNama.getText().toString();
        String identitas = mNik.getText().toString();
        String password = mPassword.getText().toString();
        if(nama.length() == 0 || identitas.length() == 0){
            tampilPesan("Silahkan lengkapi form!");
        } else {
            showDialog();
           try{
               Call<AkunModel> call = APIService.Factory.create(getContext()).updateAkun(idUser, nama,identitas,password);
               call.enqueue(new Callback<AkunModel>() {
                   @Override
                   public void onResponse(Call<AkunModel> call, Response<AkunModel> response) {
                       if(response.isSuccessful()) {
                           hideDialog();
                           if (response.body() != null) {
                             loadData(idUser);
                             tampilPesan("Data Akun berhasil diperbaharui!");
                           }
                       } else {
                           hideDialog();
                           ErrorModel error = ErrorHelper.parseError(response);
                           tampilPesan(error.message());
                       }
                   }

                   @Override
                   public void onFailure(Call<AkunModel> call, Throwable t) {
                       hideDialog();
                       tampilPesan(t.getMessage());
                   }
               });
           }catch (Exception e){
               hideDialog();
               e.printStackTrace();
               tampilPesan(e.getMessage());
           }
        }
    }


    public void loadData(String idUser)
    {
        showDialog();
        try{
            Call<AkunModel> call = APIService.Factory.create(getContext()).tampilAKun(idUser);
            call.enqueue(new Callback<AkunModel>() {
                @EverythingIsNonNull
                @Override
                public void onResponse(Call<AkunModel> call, Response<AkunModel> response) {
                    if(response.isSuccessful()) {
                        hideDialog();
                        if (response.body() != null) {
                            mEmail.setText(response.body().getEmail());
                            mNama.setText(response.body().getNama());
                            mNik.setText(response.body().getIdentitas());
                        }
                    } else {
                        hideDialog();
                        ErrorModel error = ErrorHelper.parseError(response);
                        tampilPesan(error.message());
                    }
                }
                @EverythingIsNonNull
                @Override
                public void onFailure(Call<AkunModel> call, Throwable t) {
                    hideDialog();
                    tampilPesan(t.getMessage());
                }
            });
        } catch (Exception e){
            hideDialog();
            e.printStackTrace();
            tampilPesan(e.getMessage());
        }
    }


    public void setData(View view){
        mEmail = view.findViewById(R.id.txtEmail);
        mEmail.setEnabled(false);
        mNama = view.findViewById(R.id.txtNama);
        mNik = view.findViewById(R.id.txtIdentitas);
        mPassword = view.findViewById(R.id.txtPassword);
        mEdit = view.findViewById(R.id.btnEdit);
        mLogout = view.findViewById(R.id.btnLogout);
        pDialog = new ProgressDialog(getContext());
        pDialog.setCancelable(false);
        pDialog.setMessage("Loading.....");
    }
    private void showDialog(){
        if(!pDialog.isShowing()){
            pDialog.show();
        }
    }

    private void hideDialog(){
        if(pDialog.isShowing()){
            pDialog.dismiss();
        }
    }

    public void tampilPesan(String msg)
    {
        Toast.makeText(getContext(), msg, Toast.LENGTH_LONG).show();
    }

}

File: android/facevoting/app/src/main/java/com/berkatfaatulohalawa1711010164/facevoting/fragment/akun_fragment.kt
package com.berkatfaatulohalawa1711010164.facevoting.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.berkatfaatulohalawa1711010164.facevoting.R
import com.berkatfaatulohalawa1711010164.facevoting.core.Constants
import com.berkatfaatulohalawa1711010164.facevoting.core.Resource
import com.berkatfaatulohalawa1711010164.facevoting.core.SessionHelper
import com.berkatfaatulohalawa1711010164.facevoting.ui.Login
import com.berkatfaatulohalawa1711010164.facevoting.viewmodel.AkunViewModel
import android.content.Context

class akun_fragment : Fragment() {
    private lateinit var pDialog: AlertDialog
    private lateinit var mEmail: EditText
    private lateinit var mNama: EditText
    private lateinit var mNik: EditText
    private lateinit var mPassword: EditText
    private lateinit var toolbar: Toolbar
    private lateinit var mEdit: Button
    private lateinit var mLogout: Button

    private val viewModel: AkunViewModel by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val myview = inflater.inflate(R.layout.fragment_akun, container, false)
        setData(myview)
        activity?.let {
            val act = it as AppCompatActivity
            act.setSupportActionBar(toolbar)
            toolbar.title = "Setting Akun"
        }
        val idUser = requireActivity().getSharedPreferences(Constants.USER_KEY, Context.MODE_PRIVATE)
            .getString("id_user", "") ?: ""
        viewModel.loadAkun(idUser)

        viewModel.akunState.observe(viewLifecycleOwner) { resource ->
            when (resource) {
                is Resource.Loading -> showDialog()
                is Resource.Success -> {
                    hideDialog()
                    mEmail.setText(resource.data.email)
                    mNama.setText(resource.data.nama)
                    mNik.setText(resource.data.identitas)
                }
                is Resource.Error -> {
                    hideDialog()
                    showToast(resource.message)
                }
            }
        }

        viewModel.updateState.observe(viewLifecycleOwner) { resource ->
            when (resource) {
                is Resource.Loading -> showDialog()
                is Resource.Success -> {
                    hideDialog()
                    viewModel.loadAkun(idUser)
                    showToast("Data Akun berhasil diperbaharui!")
                }
                is Resource.Error -> {
                    hideDialog()
                    showToast(resource.message)
                }
            }
        }

        mEdit.setOnClickListener {
            val nama = mNama.text.toString()
            val identitas = mNik.text.toString()
            val password = mPassword.text.toString()
            if (nama.isEmpty() || identitas.isEmpty()) {
                showToast("Silahkan lengkapi form!")
            } else {
                viewModel.updateAkun(idUser, nama, identitas, password)
            }
        }
        mLogout.setOnClickListener {
            SessionHelper.logout(context)
            startActivity(Intent(activity, Login::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            })
            activity?.finish()
        }
        return myview
    }

    fun setData(view: View) {
        toolbar = view.findViewById(R.id.toolbarAkun)
        mEmail = view.findViewById<EditText>(R.id.txtEmail).also { it.isEnabled = false }
        mNama = view.findViewById(R.id.txtNama)
        mNik = view.findViewById(R.id.txtIdentitas)
        mPassword = view.findViewById(R.id.txtPassword)
        mEdit = view.findViewById(R.id.btnEdit)
        mLogout = view.findViewById(R.id.btnLogout)
        pDialog = AlertDialog.Builder(requireContext())
            .setMessage("Loading.....")
            .setCancelable(false)
            .create()
    }

    private fun showDialog() { if (!pDialog.isShowing) pDialog.show() }
    private fun hideDialog() { if (pDialog.isShowing) pDialog.dismiss() }
    private fun showToast(msg: String) { Toast.makeText(context, msg, Toast.LENGTH_LONG).show() }
}
